File: Java/FileData.java
import jdk.nashorn.api.scripting.JSObject;

import javax.script.Invocable;
import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.script.ScriptException;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;

class FileData {
    private static class HeaderData {
        public final int rawLen = 48; // always 48 bytes
        public int coreLen;
        public int corePaddingLen;
        public int ciipherPaddingLen;
    }

    private String encryptedFileKey;
    private String baseIVec;
    private String encryptedFilePath;
    private int blockSize;
    private final HeaderData headerData;
    private String outputFilePath;
    private final byte[] supportedFileVersion = {(byte) 98, (byte) 99, (byte) 48, (byte) 49};

    // appends a incrementing number to either the output path
    // or the original path if no output was given until
    // a path is found for which no file exists yet
    // CAUTION: this can break if file names are too long
    private String checkOutputFilePath(final String currentPath) {
        int postFix = 1;
        String newPath = currentPath;
        String originalPath = currentPath;
        boolean suitablePathFound = false;

        while (!suitablePathFound) {
            if (newPath.length() == 0) {
                System.out.println("Output filepath is empty, deriving it from input");

                // first, get rid of the .bc extension
                int startPos = 0;
                int endPos = this.encryptedFilePath.lastIndexOf(".");
                if (endPos == -1) {
                    break;
                }
                newPath = originalPath = this.encryptedFilePath.substring(startPos, endPos);
            }

            if (Files.exists(Paths.get(newPath))) {
                System.out.println("Output filepath '" + newPath + "' already exists, deriving a new one");

                // insert a number after the file name
                int extensionPos = originalPath.lastIndexOf(".");
                if (extensionPos == -1) {
                    break;
                }

                newPath = originalPath.substring(0, extensionPos) + " (" + String.valueOf(postFix) + ")" + originalPath.substring(extensionPos);
                ++postFix;
            } else {
                suitablePathFound = true;
                break;
            }

            System.out.println("New output filepath: " + newPath);
        }

        if (!suitablePathFound) {
            throw new RuntimeException("Could not find a usable output filepath");
        }

        return newPath;
    }

    public FileData() {
        this.headerData = new HeaderData();
    }

    // this method should ideally be implemented using a proper JSON library,
    // but for the purpose of demonstrating which infos are needed from
    // the file header the built in scripting solution should be sufficient
    public void parseHeader(final String encryptedFilePath, final String outputFilePath) {
        System.out.println("Parsing header of encrypted file: '" + encryptedFilePath + "'");

        if (!encryptedFilePath.substring(encryptedFilePath.length() - 3).equals(".bc")) {
            throw new RuntimeException("Given filepath does not have the right extension ('.bc'), please specify a Boxcryptor encrypted file");
        }

        if (!Files.isRegularFile(Paths.get(encryptedFilePath))) {
            throw new RuntimeException("Encrypted file (" + encryptedFilePath + ") can't be opened (make sure the provided path "
                    + "is correct, the file exists and you have the right to open the file)");
        }

        this.encryptedFilePath = encryptedFilePath;

        try {
            // read the first 16 bytes which contain the file version
            // and information about the length of the different file parts
            byte[] rawHeaderBytes = new byte[16];
            InputStream is = new FileInputStream(encryptedFilePath);
            is.read(rawHeaderBytes, 0, 16);

            byte[] fileVersionBytes = Arrays.copyOfRange(rawHeaderBytes, 0, 4);
            if (!Arrays.equals(fileVersionBytes, this.supportedFileVersion)) {
                throw new RuntimeException("Unknown file version found in header, aborting...");
            }

            byte[] headerCoreLenBytes = Arrays.copyOfRange(rawHeaderBytes, 4, 8);
            byte[] headerPaddingLenBytes = Arrays.copyOfRange(rawHeaderBytes, 8, 12);
            byte[] cipherPaddingLenBytes = Arrays.copyOfRange(rawHeaderBytes, 12, 16);

            int headerCoreLen = (headerCoreLenBytes[3] & 0xFF) << 24 | (headerCoreLenBytes[2] & 0xFF) << 16
                    | (headerCoreLenBytes[1] & 0xFF) << 8 | (headerCoreLenBytes[0] & 0xFF);
            int headerPaddingLen = (headerPaddingLenBytes[3] & 0xFF) << 24 | (headerPaddingLenBytes[2] & 0xFF) << 16
                    | (headerPaddingLenBytes[1] & 0xFF) << 8 | (headerPaddingLenBytes[0] & 0xFF);
            int cipherPaddingLen = (cipherPaddingLenBytes[3] & 0xFF) << 24 | (cipherPaddingLenBytes[2] & 0xFF) << 16
                    | (cipherPaddingLenBytes[1] & 0xFF) << 8 | (cipherPaddingLenBytes[0] & 0xFF);

            this.headerData.coreLen = headerCoreLen;
            this.headerData.corePaddingLen = headerPaddingLen;
            this.headerData.ciipherPaddingLen = cipherPaddingLen;

            // go to the end of the raw header and read the core header
            is.skip(this.headerData.rawLen - 16);
            byte[] coreHeaderBytes = new byte[this.headerData.coreLen];
            is.read(coreHeaderBytes, 0, this.headerData.coreLen);

            // parse the json data in core file header
            String EXTRACTOR_SCRIPT = "var parseJSON = function(raw) { " +
                    "var json = JSON.parse(raw); " +
                    "return { blockSize: json.cipher.blockSize, " +
                    "ivec: json.cipher.iv, fileKey: json.encryptedFileKeys[0].value};};";

            ScriptEngine engine = new ScriptEngineManager().getEngineByName("nashorn");
            engine.eval(EXTRACTOR_SCRIPT);
            Invocable invocable = (Invocable) engine;
            JSObject result = (JSObject) invocable.invokeFunction("parseJSON", new String(coreHeaderBytes));

            this.blockSize = Integer.valueOf(result.getMember("blockSize").toString());
            this.baseIVec = result.getMember("ivec").toString();
            this.encryptedFileKey = result.getMember("fileKey").toString();

            this.outputFilePath = this.checkOutputFilePath(outputFilePath);
        } catch (IOException |ScriptException |NoSuchMethodException e) {
            throw new RuntimeException("Header could not be parsed", e);
        }

        System.out.println("Parsing finished");
    }

    public String getOutputFilePath() {
        return this.outputFilePath;
    }

    public String getEncryptedFileKey() {
        return this.encryptedFileKey;
    }

    public String getEncryptedFilePath() {
        return this.encryptedFilePath;
    }

    public String getBaseIVec() {
        return this.baseIVec;
    }

    public int getBlockSize() {
        return this.blockSize;
    }

    public int getHeaderLen() {
        return this.headerData.rawLen + this.headerData.coreLen + this.headerData.corePaddingLen;
    }

    public int getCipherPadding() {
        return this.headerData.ciipherPaddingLen;
    }
}

File: C++/FileData.cpp
#include "FileData.h"
#include <fstream>
#include <iostream>
#include <vector>
#include <stdexcept>
#include "TypeDefs.h"

// this method should ideally be implemented using a proper JSON library,
// but for the purpose of demonstrating which infos are needed from
// the file header simple string searches should be sufficient
bool FileData::ParseHeader(const std::string& encryptedFilePath, const std::string& outputFilePath)
{
	std::cout << "Parsing header of encrypted file: '" << encryptedFilePath << "'" << std::endl;

	if (encryptedFilePath.substr(encryptedFilePath.length() - 3) != ".bc")
	{
		throw std::runtime_error("Given filepath does not have the right extension ('.bc'), please specify a Boxcryptor encrypted file");
	}

	std::ifstream headerFile(encryptedFilePath, std::ios::binary);
	if (!headerFile.good())
	{
		std::string errorMsg("Encrypted file (" + encryptedFilePath + ") can't be opened (make sure the provided path is correct, the file exists and you have the right to open the file)");
		throw std::runtime_error(errorMsg.c_str());
	}

	this->m_encryptedFilePath = encryptedFilePath;

	// read the first 16 bytes which contain the file version
	// and information about the length of the different file parts
	std::vector<char> rawHeaderBytes(16);
	headerFile.read(&rawHeaderBytes[0], 16);

	auto fileVersionBytes = std::vector<byte>(rawHeaderBytes.begin(), rawHeaderBytes.begin() + 4);
	if (fileVersionBytes != this->m_supportedFileVersion)
	{
		throw std::runtime_error("Unknown file version found in header, aborting...");
	}

	auto headerCoreLenBytes = std::vector<byte>(rawHeaderBytes.begin() + 4, rawHeaderBytes.begin() + 8);
	auto headerPaddingLenBytes = std::vector<byte>(rawHeaderBytes.begin() + 8, rawHeaderBytes.begin() + 12);
	auto cipherPaddingLenBytes = std::vector<byte>(rawHeaderBytes.begin() + 12, rawHeaderBytes.begin() + 16);

	unsigned int headerRawLen = 48; // always 48 bytes
	unsigned int headerCoreLen = headerCoreLenBytes.at(3) << 24 | headerCoreLenBytes.at(2) << 16 | headerCoreLenBytes.at(1) << 8 | headerCoreLenBytes.at(0);
	unsigned int headerPaddingLen = headerPaddingLenBytes.at(3) << 24 | headerPaddingLenBytes.at(2) << 16 | headerPaddingLenBytes.at(1) << 8 | headerPaddingLenBytes.at(0);
	unsigned int cipherPaddingLen = cipherPaddingLenBytes.at(3) << 24 | cipherPaddingLenBytes.at(2) << 16 | cipherPaddingLenBytes.at(1) << 8 | cipherPaddingLenBytes.at(0);

	this->m_headerData.rawLen = headerRawLen;
	this->m_headerData.coreLen = headerCoreLen;
	this->m_headerData.corePaddingLen = headerPaddingLen;
	this->m_headerData.cipherPaddingLen = cipherPaddingLen;

	// go to the end of the raw header and read the core header
	headerFile.seekg(this->m_headerData.rawLen);
	std::vector<char> coreHeaderBytes(headerCoreLen);
	headerFile.read(&coreHeaderBytes[0], this->m_headerData.coreLen);

	std::string coreHeader(coreHeaderBytes.begin(), coreHeaderBytes.end());
	
	// find the blocksize
	std::string searchString = R"("blockSize")";
	size_t posBegin = coreHeader.find(searchString);
	posBegin = coreHeader.find(":", posBegin + searchString.length());
	size_t posEnd = coreHeader.find(",", posBegin + 1);

	if (posBegin != std::string::npos && posEnd != std::string::npos && posEnd - posBegin > 0)
	{
		std::string blockSize = coreHeader.substr(posBegin + 1, posEnd - posBegin - 1);
		try { this->m_blockSize = std::stoi(blockSize); }
		catch (...) { throw std::runtime_error("Could not convert block size to integer"); }
	}
	else
	{
		throw std::runtime_error("Could not find block size in file header");
	}

	// find the initialization vector
	searchString = R"("iv")";
	posBegin = coreHeader.find(searchString);
	posBegin = coreHeader.find(R"(")", posBegin + searchString.length());
	posEnd = coreHeader.find(R"(")", posBegin + 1);

	if (posBegin != std::string::npos && posEnd != std::string::npos && posEnd - posBegin > 0)
	{
		this->m_baseIVec = coreHeader.substr(posBegin + 1, posEnd - posBegin - 1);
	}
	else
	{
		throw std::runtime_error("Could not find initialization vector in file header");
	}

	// find the (first) encrypted file key object
	searchString = R"("encryptedFileKeys")";
	posBegin = coreHeader.find(searchString);
	posBegin = coreHeader.find("{", posBegin + searchString.length());
	posEnd = coreHeader.find("}", posBegin + 1);

	// and the 'value' within
	searchString = R"("value")";
	posBegin = coreHeader.find(searchString, posBegin + 1);
	posBegin = coreHeader.find(R"(")", posBegin + searchString.length());

	// check if we are still in the right block / JSON object
	if (posBegin > posEnd)
	{
		throw std::runtime_error("The (first) file key object has no suitable key value");
	}
	posEnd = coreHeader.find(R"(")", posBegin + 1);

	if (posBegin != std::string::npos && posEnd != std::string::npos && posEnd - posBegin > 0)
	{
		this->m_encryptedFileKey = coreHeader.substr(posBegin + 1, posEnd - posBegin - 1);
	}
	else
	{
		throw std::runtime_error("Could not find file key in file header");
	}

	this->m_outputFilePath = this->CheckOutputFilepath(outputFilePath);

	std::cout << "Parsing finished" << std::endl;
	return true;
}

std::string FileData::GetOutputFilepath() const
{
	return this->m_outputFilePath;
}

std::string FileData::GetEncryptedFileKey() const
{
	return this->m_encryptedFileKey;
}

std::string FileData::GetEncryptedFilePath() const
{
	return this->m_encryptedFilePath;
}

std::string FileData::GetBaseIVec() const
{
	return this->m_baseIVec;
}

unsigned int FileData::GetBlockSize() const
{
	return this->m_blockSize;
}

unsigned int FileData::GetHeaderLen() const
{
	return this->m_headerData.rawLen + this->m_headerData.coreLen + this->m_headerData.corePaddingLen;
}

unsigned int FileData::GetCipherPadding() const
{
	return this->m_headerData.cipherPaddingLen;
}

// appends a incrementing number to either the output path
// or the original path if no output was given until
// a path is found for which no file exists yet
// CAUTION: this can break if file names are too long
/*private*/ std::string FileData::CheckOutputFilepath(const std::string& currentPath)
{
	int postFix = 1;
	std::string newPath = currentPath;
	std::string originalPath = currentPath;
	bool suitablePathFound = false;
	while (!suitablePathFound)
	{
		if (newPath.length() == 0)
		{
			std::cout << "Output filepath is empty, deriving it from input" << std::endl;

			// first, get rid of the .bc extension
			size_t startPos = 0;
			size_t endPos = this->m_encryptedFilePath.find_last_of(".");
			if (endPos == std::string::npos) { break; }

			newPath = originalPath = this->m_encryptedFilePath.substr(startPos, endPos);
		}

		if (std::ifstream(newPath))
		{
			std::cout << "Output filepath '" << newPath << "' already exists, deriving a new one" << std::endl;

			// insert a number after the file name
			size_t extensionPos = originalPath.find_last_of(".");
			if (extensionPos == std::string::npos) { break; }

			newPath = originalPath.substr(0, extensionPos) + " (" + std::to_string(postFix) + ")" + originalPath.substr(extensionPos);
			++postFix;
		}
		else
		{
			suitablePathFound = true;
			break;
		}

		std::cout << "New output filepath: " << newPath << std::endl;
	}

	if (!suitablePathFound)
	{
		throw std::runtime_error("Could not find a usable output filepath");
	}

	return newPath;
}